provider/src/main/java/org/sunbird/keycloak/KeycloakSmsAuthenticator.java

```diff
@@ -0,0 +1,379 @@
+package org.sunbird.keycloak;
+
+import org.apache.http.util.TextUtils;
+import org.jboss.logging.Logger;
+import org.keycloak.authentication.AuthenticationFlowContext;
+import org.keycloak.authentication.AuthenticationFlowError;
+import org.keycloak.authentication.Authenticator;
+import org.keycloak.authentication.actiontoken.DefaultActionTokenKey;
+import org.keycloak.authentication.actiontoken.resetcred.ResetCredentialsActionToken;
+import org.keycloak.authentication.authenticators.browser.AbstractUsernameFormAuthenticator;
+import org.keycloak.common.util.Time;
+import org.keycloak.credential.CredentialModel;
+import org.keycloak.email.EmailException;
+import org.keycloak.email.EmailTemplateProvider;
+import org.keycloak.events.Details;
+import org.keycloak.events.Errors;
+import org.keycloak.events.EventBuilder;
+import org.keycloak.events.EventType;
+import org.keycloak.models.*;
+import org.keycloak.models.utils.FormMessage;
+import org.keycloak.services.ServicesLogger;
+import org.keycloak.services.messages.Messages;
+import org.keycloak.sessions.AuthenticationSessionModel;
+
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.UriBuilder;
+import java.util.Date;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Created by joris on 11/11/2016.
+ */
+public class KeycloakSmsAuthenticator implements Authenticator {
+
+    private static Logger logger = Logger.getLogger(KeycloakSmsAuthenticator.class);
+
+    public static final String CREDENTIAL_TYPE = "sms_validation";
+
+    private enum CODE_STATUS {
+        VALID,
+        INVALID,
+        EXPIRED
+    }
+
+
+    public void authenticate(AuthenticationFlowContext context) {
+        logger.debug("authenticate called ... context = " + context);
+        UserModel user = context.getUser();
+        AuthenticatorConfigModel config = context.getAuthenticatorConfig();
+
+        List<String> mobileNumberCreds = user.getAttribute("mobile");
+
+        String mobileNumber = null;
+        String userEmail = user.getEmail();
+
+        if (mobileNumberCreds != null && !mobileNumberCreds.isEmpty()) {
+            mobileNumber = mobileNumberCreds.get(0);
+        }
+
+        if (!TextUtils.isEmpty(userEmail)) {
+            logger.debug("Email is not null - " + userEmail);
+        }
+
+        if (mobileNumber != null) {
```

Comment:
@shriharshs suggest moving the sms sending to a separate method which can be tested independently.